### DISCUSSION TITLE
unable to add custom pipeline

### DISCUSSION THREAD

**Discussion by tushitdave** [explosion/spaCy#8550] [Help: Coding & Implementations] [answered]
Hi , This is in reference to the below code, I am unable to create custom pipeline(input and output both are presented in the code).
Input:- 
```
def add_new_ruler(skill_pattern_path):
    new_ent_ruler = EntityRuler(nlp).from_disk(skill_pattern_path)
    new_ent_ruler = nlp.create_pipe('new_ent_ruler',new_ent_ruler)
    nlp.add_pipe(new_ent_ruler, after='parser' , before ='ner')

print(nlp.pipe_names)
```
output:- 
![spacy pipeline issue](https://user-images.githubusercontent.com/52331739/123916789-abc87280-d99f-11eb-81c2-081c8346a2ee.PNG)
['tok2vec', 'tagger', 'parser', 'ner', 'attribute_ruler', 'lemmatizer']

python version 3.8.5
spacy version: 3.0.0
windows:10

**Comment by svlandeg** [comment, marked as answer]
Could you provide a small reproducible code snippet (not a screenshot) we can run, that showcases how you're calling `add_new_ruler`?

**Reply by tushitdave** [reply]
@svlandeg - I just want to add customize pipeline
If I apply below method, without creating a function- 

new_ent_ruler = EntityRuler(nlp).from_disk(skill_pattern_path)
new_ent_ruler = nlp.create_pipe('new_ent_ruler',new_ent_ruler)
nlp.add_pipe(new_ent_ruler, after='parser' , before ='ner')

I receive below error- 
ValueError: [E002] Can't find factory for 'new_ent_ruler' for language English (en). This usually happens when spaCy calls `nlp.create_pipe` with a custom component name that's not registered on the current language class. If you're using a Transformer, make sure to install 'spacy-transformers'. If you're using a custom component, make sure you've added the decorator `@Language.component` (for function components) or `@Language.factory` (for class components).

Available factories: attribute_ruler, tok2vec, merge_noun_chunks, merge_entities, merge_subtokens, token_splitter, parser, beam_parser, entity_linker, ner, beam_ner, entity_ruler, lemmatizer, tagger, morphologizer, senter, sentencizer, textcat, textcat_multilabel, en.lemmatizer

hence to add pipeline I have created, below function to add new customized pipeline called "new_ent_ruler"

def add_new_ruler(skill_pattern_path):
    new_ent_ruler = EntityRuler(nlp).from_disk(skill_pattern_path)
    new_ent_ruler = nlp.create_pipe('new_ent_ruler',new_ent_ruler)
    nlp.add_pipe(new_ent_ruler, after='parser' , before ='ner')

print(nlp.pipe_names)

Calling of this function is not required since this is a part of very algorithm function which I am still working on.

**Reply by tushitdave** [reply]
![Capture1](https://user-images.githubusercontent.com/52331739/123951945-f90a0b80-d9c2-11eb-9416-57959eac10dc.PNG)

Hi, Even If I want to apply @Language.component- Still unable to add customized pipeline.

**Reply by svlandeg** [reply]
It might be worthwhile to try the example from the docs first and see how that works: https://spacy.io/usage/processing-pipelines#custom-components

You need `@Language.component("UNIQUE_NAME")` to decorate the function that defines your custom component, and then you need to make sure to call `nlp.add_pipe("UNIQUE_NAME")`.

**Reply by adrianeboyd** [reply]
On the other hand, if you're just using an entity ruler, you shouldn't need to define a new component here, and a lot of workflows are easier if you use built-in components where possible.

```
nlp = spacy.load("en_core_web_sm")
nlp.add_pipe("entity_ruler")
nlp.get_pipe("entity_ruler").from_disk(pattern_path)
nlp.to_disk("/path/to/model") # model saved with patterns included, can be reloaded with spacy.load("/path/to/model")
```

See my answer to a related stackoverflow question: https://stackoverflow.com/a/67916773

**Reply by tushitdave** [reply]
Thanks @svlandage and adrianeboyd ... I got this

**Reply by tushitdave** [reply]
```
@Language.factory("new_ruler")
def new_ruler(skill_pattern_path):
    new_ruler = EntityRuler(nlp).from_disk(skill_pattern_path)
    return skill_pattern_path
nlp.add_pipe(new_ruler, after='parser')
print(nlp.pipe_names)
```

```
ValueError: [E004] Can't set up pipeline component: a factory for 'new_ruler' already exists. Existing factory: <function Language.component.<locals>.add_component.<locals>.factory_func at 0x000001A7F310F1F0>. New factory: <function new_ruler at 0x000001A7F3125A60>
```

This is in reference to above input and output , If the new_ruler component already been added inside the pipeline than it should reflect in the below pipeline

![image](https://user-images.githubusercontent.com/52331739/126267365-39d34be4-8a14-48fe-ae00-1a16922dc2ce.png)
